// @cashflow/consolidation-service/src/infrastructure/persistence/prisma.service.ts
import { Injectable, OnModuleInit, OnModuleDestroy } from '@nestjs/common';
import { PrismaClient } from '.prisma/client-consolidation';
import { Logger } from '@shared/infrastructure/logging/logger';

@Injectable()
export class PrismaService extends PrismaClient implements OnModuleInit, OnModuleDestroy {
  private readonly logger: Logger;

  constructor() {
    super({
      log: [
        { emit: 'event', level: 'query' },
        { emit: 'event', level: 'error' },
        { emit: 'event', level: 'info' },
        { emit: 'event', level: 'warn' },
      ],
    });

    this.logger = new Logger({ service: 'prisma-consolidation' });
  }

  async onModuleInit() {
    await this.$connect();
    this.logger.info('Prisma connected to consolidation database');

    // Log queries in development
    if (process.env.NODE_ENV === 'development') {
      this.$on('query' as never, (e: any) => {
        this.logger.debug('Query executed', {
          query: e.query,
          params: e.params,
          duration: `${e.duration}ms`,
        });
      });
    }

    this.$on('error' as never, (e: any) => {
      this.logger.error('Prisma error', e);
    });
  }

  async onModuleDestroy() {
    await this.$disconnect();
    this.logger.info('Prisma disconnected from consolidation database');
  }
}

// Made with Bob
